pkg/sac/allow_fixed_scopes_checker_core.go

```diff
@@ -3,121 +3,256 @@ package sac
+	if c.accessKeys.Cardinality() == 0 &&
+		c.resourceKeys.Cardinality() == 0 &&
+		c.clusterKeys.Cardinality() == 0 &&
+		c.resourceKeys.Cardinality() == 0 {
 		return effectiveaccessscope.UnrestrictedEffectiveAccessScope(), nil
 	}
-	_, accessAllowed := c.topLevelKeys()[AccessModeScopeKey(resource.Access)]
-	if !accessAllowed {
+	if !c.accessKeys.Contains(AccessModeScopeKey(resource.Access)) {
 		return effectiveaccessscope.DenyAllEffectiveAccessScope(), nil
 	}
-	return c.next().getResourceEffectiveAccessScope(resource)
-}
-
-func (c allowFixedScopesCheckerCore) getResourceEffectiveAccessScope(resource permissions.ResourceWithAccess) (*effectiveaccessscope.ScopeTree, error) {
-	if len(c) == 0 {
+	if c.resourceKeys.Cardinality() == 0 &&
+		c.clusterKeys.Cardinality() == 0 &&
+		c.resourceKeys.Cardinality() == 0 {
 		return effectiveaccessscope.UnrestrictedEffectiveAccessScope(), nil
 	}
-	_, resourceAllowed := c.topLevelKeys()[ResourceScopeKey(resource.Resource.GetResource())]
-	if !resourceAllowed {
-		if resource.Resource.GetReplacingResource() == nil {
-			return effectiveaccessscope.DenyAllEffectiveAccessScope(), nil
-		}
-		_, replacingResourceAllowed := c.topLevelKeys()[ResourceScopeKey(*resource.Resource.GetReplacingResource())]
-		if !replacingResourceAllowed {
-			return effectiveaccessscope.DenyAllEffectiveAccessScope(), nil
-		}
+	targetResource := resource.Resource.GetResource()
+	targetReplacingResource := resource.Resource.GetReplacingResource()
+	if !c.resourceKeys.Contains(ResourceScopeKey(targetResource)) &&
+		(targetReplacingResource == nil || !c.resourceKeys.Contains(ResourceScopeKey(*targetReplacingResource))) {
+		return effectiveaccessscope.DenyAllEffectiveAccessScope(), nil
 	}
-	return c.next().getClusterEffectiveAccessScope()
-}
-
-func (c allowFixedScopesCheckerCore) getClusterEffectiveAccessScope() (*effectiveaccessscope.ScopeTree, error) {
-	if len(c) == 0 {
+	if c.clusterKeys.Cardinality() == 0 &&
+		c.namespaceKeys.Cardinality() == 0 {
 		return effectiveaccessscope.UnrestrictedEffectiveAccessScope(), nil
 	}
-	clusterIDs := make([]string, 0, len(c[0]))
-	for clusterID := range c.topLevelKeys() {
-		clusterIDs = append(clusterIDs, clusterID.String())
+	clusterIDs := make([]string, 0, c.clusterKeys.Cardinality())
+	for _, clusterKey := range c.clusterKeys.AsSlice() {
+		clusterIDs = append(clusterIDs, clusterKey.String())
 	}
-	if len(c) == 1 {
+	if c.namespaceKeys.Cardinality() == 0 {
 		return effectiveaccessscope.FromClustersAndNamespacesMap(clusterIDs, nil), nil
 	}
-	namespaces := make([]string, 0, len(c[1]))
-	for namespace := range c.next().topLevelKeys() {
-		namespaces = append(namespaces, namespace.String())
+	namespaces := make([]string, 0, c.namespaceKeys.Cardinality())
+	for _, namespaceKey := range c.namespaceKeys.AsSlice() {
+		namespaces = append(namespaces, namespaceKey.String())
 	}
-	clusterNamespaceMap := make(map[string][]string, 0)
+	namespaceMap := make(map[string][]string, len(clusterIDs))
```

Comment:
if you keep old names then diff will be smaller
```suggestion
	clusterNamespaceMap := make(map[string][]string, len(clusterIDs))
```